### ISSUE TITLE
Multithreaded Compilation

### ISSUE THREAD

**Issue by brockelmore** [foundry-rs/foundry#166]
# Requested Feature
For sufficiently large repos, solc compilation can be the slowest part of testing. In those cases we should multithread compilation as much as possible. Here is a rough sketch of a potential way to achieve this?

## Suggested solution

1. Determine a threshold for multithreading compilation (# of sol files in repo?) that offsets any overhead
2. Spin up 1 thread per file, determining dependency graph
3. Hash trees/dep graph, and remove duplicate trees
4. Compile the needed contracts using the existing threads, in parallel

There could be issues here that I am unaware of but documenting the thoughts I had anyway

**Comment by maxsam4** [comment]
I feel like the overhead will not be worth it in most cases. Individual contracts are usually so small that invoking solc is going to consume more time than solc compiling them.

This feature would make more sense in solc itself but I won't put it on the priority list even there.

**Comment by lucas-manuel** [comment]
I tend to agree that it would be cool but could be more trouble than it is worth, especially if we have compilation caching etc.

**Comment by mattsse** [comment]
this is actually already in use for workspaces that require different solc versions
https://github.com/gakonst/ethers-rs/pull/652

**Comment by gakonst** [comment]
@mattsse bumping this, seems like next up in our priority list along with #769?

**Comment by onbjerg** [comment]
Marking this and #769 as high prio if I understand your comment correctly

**Comment by lucas-manuel** [comment]
Close-able? @mds1

**Comment by mds1** [comment]
@mattsse wdyt here? Slow compilation times are a big pain point these days so this could be valuable, and per https://github.com/foundry-rs/foundry/issues/166#issuecomment-1001475406 it sounds like it might not be too hard to implement

**Comment by brockelmore** [comment]
> I believe this is not tractable: [gakonst/ethers-rs#943 (comment)](https://github.com/gakonst/ethers-rs/pull/943#issuecomment-1253022311). What makes compilation slow these days? Repos got bigger? Via-ir? Forge-std?
> 
> (Closing)

Fwiw, i do think for some repos it may be worthwhile dependent on the number of roots in the project. Where a root is a contract in which no other contract inherits or imports it (diagram below shows how we could split a project into two roots based on dependency graph). I generally believe this is tractable but the number of projects that need this may be few and far between? i.e. if I import my complex contract A, and use the contract as an interface elsewhere i still have to compile the contract in its entirety (but if the project uses interface contracts its probably worth splitting up). This is a different version of parallelization than trying to do any crazy linking. It is much more straightforward

```mermaid
graph TD;
    subgraph root_2 
         A-->B;
         A-->C;
         C-->G;
         B-->H;
    end
    subgraph root_1
        D-->E;
        D-->F;
    end
```

**Comment by 0xLightt** [comment]
Hi all,

I'm revisiting the idea of parallel compilation in Foundry, particularly in light of the Solidity team's current stance against implementing this in `solc`. With the shift towards `via-ir` potentially becoming the default mode for Solidity compilation, I believe this feature could become even more crucial for optimizing build times, especially for larger projects.

Since this issue was previously closed, has there been any new developments or considerations about integrating parallel compilation in Foundry? I recognize the complexity of this feature but am interested in whether there's room for further discussion or contributions.

Thanks for your work on Foundry!

**Comment by mattsse** [comment]
no plans atm because the complexity overhead is quite huge.

but I can see that this would still be a great feature to have so reopening to not lose it

**Comment by k06a** [comment]
@gakonst option to compile files in parallel could be great and relatively easy solution.

**Comment by dezzeus** [comment]
Any marginal (even experimental opt-in) improvement would be very appreciated as my team and I are working on a major project with around 250 solidity files and it takes tens of minutes to compile…

**Comment by cameel** [comment]
Hey everyone! I've been tasked with exploring this topic in the context of improving the speed of compilation via IR. We wanted to know what the obstacles for this kind of parallelized compilation are and if we can do anything to make it easier for frameworks.

Theoretically, the parallelization is very simple already: take the Standard JSON input containing all the sources and split it into series of inputs where each one uses `settings.outputSelection` to request output only for a single contract. The compiler will perform compilation and optimization only for the one you selected. It will still analyze all the sources, but the later stages of the pipeline are orders of magnitude slower than analysis so it should not matter *that* much.

To benchmark it, I created a proof of concept script ([parasolc](https://github.com/cameel/parasolc)) that can be passed in place of a `solc` binary to `forge --use`. Here's also the full report with my findings: [The parasolc experiment](https://notes.ethereum.org/@solidity/parasolc-experiment).

Now, as some have already suspected here, the overhead of doing it this way is just staggering, but now we have actual numbers. The projects I benchmarked require 3-4 times as much work compared to sequential compilation. The report attempts to explain where all that time is going, but the short of it is that this kind of parallelization makes the compiler repeat the same work multiple times in several ways (that may be different for different projects). Bytecode dependencies (contracts deployed with `new`) can no longer be reused for contracts that depend on them and also the same sources are analyzed multiple times.

![uniswap](https://github.com/foundry-rs/foundry/assets/137030/f217528f-01d8-4d34-8ec0-b3fb45a750e9)
![openzeppelin](https://github.com/foundry-rs/foundry/assets/137030/a9661743-fc43-42c1-9058-6a3a9e211bf8)

Still, while expensive, this method does provide an actual improvement in terms of wall-clock-time spent on compilation. Here are the numbers are for an 8 core machine. 

| Benchmark                             | Real time | CPU time |
|---------------------------------------|----------:|---------:|
| OpenZeppelin (sequential)             |      39 s |     39 s |
| OpenZeppelin (parallelized)           |      27 s |    144 s |
| Uniswap v4 (sequential)               |     166 s |    165 s |
| Uniswap v4 (parallelized)             |      76 s |    499 s |

It appears that with enough cores you can still come out ahead despite the overhead. While this is far from what I was hoping to present you here, and does not seem like a good choice for the default compilation mode, it's still a trade-off that may make sense in some situations. It may work better for some projects than others, depending on how they are structured and how interdependent their contracts are. The method is simple enough that it might make sense to be an optional feature.

### Can we do better?
I explored workarounds, like grouping bytecode dependencies together to improve reuse or culling of sources irrelevant to the contract being compiled. That's also in the report. Both unfortunately come with significant downsides and/or don't improve the situation as much as one would hope.

There's the idea described by @brockelmore above (and already used by e.g. Hardhat) of identifying independent clusters of contracts and compiling each cluster separately. It's largely orthogonal to what I explored here - it just shifts the problem inside the groups - and it's not effective against projects with tightly interconnected sources. Still, I think it's worthwhile when applicable, and we'd like to see more tools using it. It is not technically complicated and all the information is already out there, but it does require parsing the AST to identify the imports. One thing we could do to make it more straightforward to do would be to make the dependency graph between sources available as a separate output. Or even just outright assign each contract to a group based on it. Would Foundry make use of such an output?

Another idea, that would only work for IR though, would be to do unoptimized IR generation sequentially and then only do optimization in parallel by compiling the produced Yul. Code generation is slower than analysis, but it's the optimization and Yul->EVM transform that are the real bottlenecks so it should provide some savings. The downside here is that it would be a little more complex than the naive method and only viable after we fix https://github.com/ethereum/solidity/issues/15062. It also only addresses the analysis overhead and would do nothing for bytecode dependencies.

In the longer term, there are things we could improve in the compiler, but I can't really tell if/when they will happen:
- **Reuse of analysis info**: We can avoid overhead by just not doing analysis in parallel. This requires a way to reuse it - the current AST export does not cut it. The import redoes a lot of analysis.
- **Lazy analysis**: Avoid analyzing definitions that are not relevant. Currently the analysis works in distinct phases, where each has to analyze the whole input, because it's not known which parts the next steps will use.
- **Bytecode reuse**: A way to serialize and reuse already compiled bytecode for contracts that depend on it.
- **Multithreaded compilation**: TBH we consider this one too complex and disruptive to introduce in the current state of the compiler, so while it could happen some day, it won't be any time soon. We're rather explore the ideas based on multiprocessing first.

**Comment by grandizzy** [comment]
we will get this out of the box with moving on Solar, no change in foundry
